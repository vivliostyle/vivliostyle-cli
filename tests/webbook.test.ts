import './mocks/fs.js';
import './mocks/tmp.js';
import './mocks/vivliostyle__jsdom.js';

import { vol } from 'memfs';
import { format } from 'prettier';
import { beforeEach, expect, it, vi } from 'vitest';
import { VivliostyleConfigSchema } from '../src/config/schema.js';
import { runCommand, toTree } from './command-util.js';

const mockedThemeModule = vi.hoisted(() => ({
  checkThemeInstallationNecessity: vi.fn(),
  installThemeDependencies: vi.fn(),
}));

vi.mock('../src/processor/theme', () => mockedThemeModule);

beforeEach(() => vol.reset());

it('generate webpub from single markdown file', async () => {
  vol.fromJSON({
    '/work/input/foo.md': '# Hi',
  });
  await runCommand(['build', './input/foo.md', '-o', 'output'], {
    cwd: '/work',
  });

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json'] as string);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
});

it('generate webpub from vivliostyle.config.js', async () => {
  const themeDir = {
    'mytheme/package.json': JSON.stringify({
      name: 'mytheme',
      main: './%style%.css',
    }),
    'mytheme/%style%.css': '/* style */',
  };
  mockedThemeModule.checkThemeInstallationNecessity.mockImplementationOnce(
    () => true,
  );
  mockedThemeModule.installThemeDependencies.mockImplementationOnce(() => {
    vol.fromJSON(themeDir, '/work/input/.vivliostyle/themes/node_modules');
  });

  const config: VivliostyleConfigSchema = {
    entry: [
      { rel: 'cover', path: 'tmpl/cover-template.html', output: 'cover.html' },
      { rel: 'contents', path: 'tmpl/toc-template.html', output: 'index.html' },
      'doc/one.md',
      'doc/two.md',
      'doc/escape check%.md',
    ],
    output: ['/work/input/output', '/work/output'],
    toc: true,
    cover: 'cover.png',
    theme: ['../mytheme', './style sheet.css'],
    readingProgression: 'rtl',
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/doc/one.md': 'yuno',
    '/work/input/doc/two.md': 'yunocchi',
    '/work/input/doc/escape check%.md': 'txt',
    '/work/input/cover.png': '',
    '/work/input/style sheet.css': '/* style */',
    '/work/input/tmpl/cover-template.html':
      '<html><body><img role="doc-cover" data-test="true" /></body></html>',
    '/work/input/tmpl/toc-template.html':
      '<html><body><nav role="doc-toc" data-test="true"></nav></body></html>',
  });
  vol.fromJSON(themeDir, '/work');
  await runCommand(['build'], { cwd: '/work/input' });

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json']!);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();

  const manifest2 = JSON.parse(
    file['/work/input/output/publication.json'] as string,
  );
  delete manifest2.dateModified;
  expect(manifest2).toEqual(manifest);
});

it('generate webpub from a plain HTML', async () => {
  vol.fromJSON({
    '/work/input/webbook.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <link rel="stylesheet" type="text/css" href="style.css">
        <title>Doc title</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/style.css': '',
  });
  await runCommand(['build', 'input/webbook.html', '-o', 'output'], {
    cwd: '/work',
  });

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json']!);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
  const entry = file['/work/output/webbook.html']!;
  expect(await format(entry, { parser: 'html' })).toMatchSnapshot();
});

it('generate webpub from a single-document publication', async () => {
  vol.fromJSON({
    '/work/input/webbook.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document with toc</title>
        <link rel="publication" href="#wp_manifest">
        <script type="application/ld+json" id="wp_manifest">
          ${JSON.stringify({
            '@context': [
              'https://schema.org',
              'https://www.w3.org/ns/wp-context',
            ],
            conformsTo: 'yuno',
            resources: ['assets/%E6%97%A5%E6%9C%AC%E8%AA%9E.svg'],
            readingOrder: [
              '#foo',
              '../bar.html',
              'subdir/index.html',
              'escape%20check%25.html',
            ],
          })}
        </script>
      </head>
      <body>

      </body>
      </html>
    `,
    '/work/input/subdir/index.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <title>Doc title</title>
        <link rel="stylesheet" href="../assets/subdir.css">
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/escape check%.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <title>Escape check</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/work/input/assets/日本語.svg': '<svg></svg>',
    '/work/input/assets/subdir.css': '',
  });
  await runCommand(['build', 'input/webbook.html', '-o', 'output'], {
    cwd: '/work',
  });

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const entry = file['/work/output/webbook.html']!;
  expect(await format(entry, { parser: 'html' })).toMatchSnapshot();
  expect(file['/work/output/escape check%.html']).toBe(
    file['/work/input/escape check%.html'],
  );
  expect(file['/work/output/assets/日本語.svg']).toBe(
    file['/work/input/assets/日本語.svg'],
  );
});

it('generate webpub from remote HTML documents with publication manifest', async () => {
  vol.fromJSON({
    '/remote/dir/index.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document</title>
        <link rel="publication" href="../publication.json">
      </head>
      <body>
      </body>
      </html>
    `,
    '/remote/dir/escape check%.html': /* html */ `
      <!DOCTYPE html>
      <html lang="ja">
      <head>
        <title>Escape check</title>
      </head>
      <body>
      </body>
      </html>
    `,
    '/remote/publication.json': JSON.stringify({
      '@context': ['https://schema.org', 'https://www.w3.org/ns/pub-context'],
      conformsTo: 'yuno',
      resources: ['../assets/%E6%97%A5%E6%9C%AC%E8%AA%9E.png'],
      readingOrder: ['dir/index.html', 'dir/escape%20check%25.html'],
    }),
    '/assets/日本語.png': 'image',
  });
  await runCommand(
    ['build', 'https://example.com/remote/dir', '-o', 'output'],
    { cwd: '/work' },
  );
  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  expect(file['/work/output/remote/dir/index.html']).toBe(
    file['/remote/dir/index.html'],
  );
  expect(file['/work/output/remote/publication.json']).toBe(
    file['/remote/publication.json'],
  );
  expect(file['/work/output/remote/dir/escape check%.html']).toBe(
    file['/remote/dir/escape check%.html'],
  );
  expect(file['/work/output/assets/日本語.png']).toBe(
    file['/assets/日本語.png'],
  );
});

it('generate webpub from a remote HTML document', async () => {
  vol.fromJSON({
    '/remote/foo bar%/escape check%.html': /* html */ `
      <html lang="en">
      <head>
        <title>Document</title>
        <link rel="stylesheet" type="text/css" href="../%E3%81%82/%E6%97%A5%E6%9C%AC%E8%AA%9E.css">
      </head>
      <body>
      </body>
      </html>
    `,
    '/remote/あ/日本語.css': '/* css */',
  });
  await runCommand(
    [
      'build',
      'https://example.com/remote/foo%20bar%25/escape%20check%25.html',
      '-o',
      'output',
    ],
    { cwd: '/work' },
  );
  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const manifest = JSON.parse(file['/work/output/publication.json']!);
  delete manifest.dateModified;
  expect(manifest).toMatchSnapshot();
  const entry = file['/work/output/remote/foo bar%/escape check%.html']!;
  expect(await format(entry, { parser: 'html' })).toMatchSnapshot();
  expect(file['/work/output/remote/あ/日本語.css']).toBe(
    file['/remote/あ/日本語.css'],
  );
});

it('generate webpub from a data URI input', async () => {
  const html = /* html */ `<!DOCTYPE html><html lang="en"><head><meta charset="UTF-8"><title></title></head><body>Hi</body></html>`;
  await runCommand(
    ['build', `data:text/html;charset=utf-8,${html}`, '-o', 'output'],
    { cwd: '/work' },
  );
  await runCommand(
    [
      'build',
      `data:text/html;base64,${Buffer.from(html).toString('base64')}`,
      '-o',
      'output2',
    ],
    { cwd: '/work' },
  );

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
  const file = vol.toJSON();
  const entry = file['/work/output/index.html']!;
  const outHtml = await format(entry, { parser: 'html' });
  expect(outHtml).toMatchSnapshot();
  expect(
    await format(file['/work/output2/index.html']!, { parser: 'html' }),
  ).toBe(outHtml);
});

it('generate webpub with complex copyAsset settings', async () => {
  const config: VivliostyleConfigSchema = {
    copyAsset: {
      includes: ['node_modules/pkgB'],
      excludes: ['cover.png'],
      includeFileExtensions: ['jxl'],
      excludeFileExtensions: ['svg'],
    },
    entry: ['doc.md'],
    output: ['/work/output'],
    workspaceDir: '.vs',
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/package.json': '',
    '/work/input/doc.md': 'yuno',
    '/work/input/cover.png': '',
    '/work/input/assetA.jxl': '',
    '/work/input/assetB.svg': '',
    '/work/input/node_modules/pkgA/img.png': '',
    '/work/input/node_modules/pkgB/a.html': '',
    '/work/input/node_modules/pkgB/bar/b.html': '',
  });
  await runCommand(['build'], { cwd: '/work/input' });

  expect(toTree(vol, { dir: '/work/output' })).toMatchSnapshot();
});

it('copy webpub assets properly', async () => {
  const themeDir = {
    'themeA/package.json': '{"main": "theme.css"}',
    'themeA/theme.css': '',
    'themeA/example/a.css': '',
    '@org/themeB/package.json': '{"main": "theme.css"}',
    '@org/themeB/theme.css': '',
    '@org/themeB/example/a.css': '',
  };
  mockedThemeModule.checkThemeInstallationNecessity.mockImplementationOnce(
    () => true,
  );
  mockedThemeModule.installThemeDependencies.mockImplementationOnce(() => {
    vol.fromJSON(themeDir, '/work/input/.vivliostyle/themes/node_modules');
  });

  const config: VivliostyleConfigSchema = {
    entry: ['doc.md'],
    output: ['/work/input/output'],
    theme: ['themeA', '@org/themeB'],
  };
  vol.fromJSON({
    '/work/input/vivliostyle.config.json': JSON.stringify(config),
    '/work/input/package.json': '',
    '/work/input/doc.md': 'yuno',
    '/work/input/node_modules/pkgA/a.html': '',
    '/work/input/node_modules/pkgA/a.css': '',
  });
  await runCommand(['build'], { cwd: '/work/input' });

  expect(toTree(vol, { dir: '/work/input/output' })).toMatchSnapshot();
});
