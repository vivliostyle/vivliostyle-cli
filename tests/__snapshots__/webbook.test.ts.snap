// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copy webpub assets properly 1`] = `
"output/
├─ doc.html
├─ publication.json
└─ themes/
   └─ node_modules/
      ├─ @org/
      │  └─ themeB/
      │     └─ theme.css
      └─ themeA/
         └─ theme.css"
`;

exports[`generate webpub from a data URI input 1`] = `
"output/
├─ index.html
└─ publication.json"
`;

exports[`generate webpub from a data URI input 2`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <title></title>
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"publication.json\\"
    />
  </head>
  <body>
    Hi
  </body>
</html>
"
`;

exports[`generate webpub from a plain HTML 1`] = `
"output/
├─ publication.json
├─ style.css
└─ webbook.html"
`;

exports[`generate webpub from a plain HTML 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "inLanguage": "ja",
  "links": Array [],
  "name": "Doc title",
  "readingOrder": Array [
    Object {
      "url": "webbook.html",
    },
  ],
  "resources": Array [
    "style.css",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from a plain HTML 3`] = `
"<!DOCTYPE html>
<html lang=\\"ja\\">
  <head>
    <link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"style.css\\" />
    <title>Doc title</title>
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"publication.json\\"
    />
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from a remote HTML document 1`] = `
"output/
├─ publication.json
└─ remote/
   ├─ foo bar%/
   │  └─ escape check%.html
   └─ あ/
      └─ 日本語.css"
`;

exports[`generate webpub from a remote HTML document 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "inLanguage": "en",
  "links": Array [],
  "name": "Document",
  "readingOrder": Array [
    Object {
      "url": "remote/foo%20bar%25/escape%20check%25.html",
    },
  ],
  "resources": Array [
    "remote/%E3%81%82/%E6%97%A5%E6%9C%AC%E8%AA%9E.css",
  ],
  "type": "Book",
}
`;

exports[`generate webpub from a remote HTML document 3`] = `
"<html lang=\\"en\\">
  <head>
    <title>Document</title>
    <link
      rel=\\"stylesheet\\"
      type=\\"text/css\\"
      href=\\"../%E3%81%82/%E6%97%A5%E6%9C%AC%E8%AA%9E.css\\"
    />
    <link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"../../publication.json\\"
    />
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from a single-document publication 1`] = `
"output/
├─ assets/
│  ├─ subdir.css
│  └─ 日本語.svg
├─ escape check%.html
├─ subdir/
│  └─ index.html
└─ webbook.html"
`;

exports[`generate webpub from a single-document publication 2`] = `
"<html lang=\\"en\\">
  <head>
    <title>Document with toc</title>
    <link rel=\\"publication\\" href=\\"#wp_manifest\\" />
    <script type=\\"application/ld+json\\" id=\\"wp_manifest\\">
      {
        \\"@context\\": [\\"https://schema.org\\", \\"https://www.w3.org/ns/wp-context\\"],
        \\"conformsTo\\": \\"yuno\\",
        \\"resources\\": [\\"assets/%E6%97%A5%E6%9C%AC%E8%AA%9E.svg\\"],
        \\"readingOrder\\": [
          \\"#foo\\",
          \\"../bar.html\\",
          \\"subdir/index.html\\",
          \\"escape%20check%25.html\\"
        ]
      }
    </script>
  </head>
  <body></body>
</html>
"
`;

exports[`generate webpub from remote HTML documents with publication manifest 1`] = `
"output/
├─ assets/
│  └─ 日本語.png
└─ remote/
   ├─ dir/
   │  ├─ escape check%.html
   │  └─ index.html
   └─ publication.json"
`;

exports[`generate webpub from single markdown file 1`] = `
"output/
├─ foo.html
└─ publication.json"
`;

exports[`generate webpub from single markdown file 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "links": Array [],
  "name": "Hi",
  "readingOrder": Array [
    Object {
      "name": "Hi",
      "url": "foo.html",
    },
  ],
  "resources": Array [],
  "type": "Book",
}
`;

exports[`generate webpub from vivliostyle.config.js 1`] = `
"output/
├─ cover.html
├─ cover.png
├─ doc/
│  ├─ escape check%.html
│  ├─ one.html
│  └─ two.html
├─ index.html
├─ publication.json
├─ style sheet.css
└─ themes/
   └─ node_modules/
      └─ mytheme/
         └─ %style%.css"
`;

exports[`generate webpub from vivliostyle.config.js 2`] = `
Object {
  "@context": Array [
    "https://schema.org",
    "https://www.w3.org/ns/pub-context",
  ],
  "conformsTo": "https://github.com/vivliostyle/vivliostyle-cli",
  "links": Array [],
  "name": "output",
  "readingOrder": Array [
    Object {
      "rel": "cover",
      "type": "LinkedResource",
      "url": "cover.html",
    },
    Object {
      "name": "Table of Contents",
      "rel": "contents",
      "type": "LinkedResource",
      "url": "index.html",
    },
    Object {
      "url": "doc/one.html",
    },
    Object {
      "url": "doc/two.html",
    },
    Object {
      "url": "doc/escape%20check%25.html",
    },
  ],
  "readingProgression": "rtl",
  "resources": Array [
    Object {
      "encodingFormat": "image/png",
      "name": "Cover image",
      "rel": "cover",
      "url": "cover.png",
    },
    "style%20sheet.css",
    "themes/node_modules/mytheme/%25style%25.css",
  ],
  "type": "Book",
}
`;

exports[`generate webpub with complex copyAsset settings 1`] = `
"output/
├─ assetA.jxl
├─ doc.html
├─ node_modules/
│  └─ pkgB/
│     ├─ a.html
│     └─ bar/
│        └─ b.html
└─ publication.json"
`;
